File: packages/input-bindings/src/repair.ts
import {
  type Binding,
  type Conflict,
  type ConflictKind,
  type WhenExpr,
} from "./index.ts";

export type ConflictDisposition =
  | "redundant"
  | "ambiguous"
  | "orderedOverride"
  | "chordPrefix"
  | "potential";

export type ConflictRepairKeepReason =
  | "existingPrecedence"
  | "potentialConflict"
  | "redundantSameAction";

export type ConflictRepair =
  | { kind: "keep"; reason: ConflictRepairKeepReason }
  | { kind: "unbind"; bindingId: string }
  | { kind: "prefer"; bindingId: string; overBindingId: string; priority: number }
  | {
      kind: "narrowContext";
      bindingId: string;
      againstBindingId: string;
      when: WhenExpr;
    };

export interface ConflictRepairPlan {
  conflict: Conflict;
  disposition: ConflictDisposition;
  repairs: ConflictRepair[];
}

const MAX_PRIORITY = 2_147_483_647;
const ALWAYS: WhenExpr = { op: "always" };

/**
 * Returns deterministic repair alternatives without applying any of them.
 * The caller remains responsible for choosing a repair and persisting the resulting profile delta.
 */
export function planConflictRepairs(
  bindings: readonly Binding[],
  conflict: Conflict,
): ConflictRepairPlan {
  const left = bindings.find((binding) => binding.id === conflict.leftBindingId);
  const right = bindings.find((binding) => binding.id === conflict.rightBindingId);
  const disposition = dispositionForKind(conflict.kind);
  if (!left || !right) {
    return { conflict: structuredClone(conflict), disposition, repairs: [] };
  }

  const repairs: ConflictRepair[] = [];

  switch (conflict.kind) {
    case "duplicate":
      repairs.push({ kind: "keep", reason: "redundantSameAction" });
      break;
    case "overrideExact":
      repairs.push({ kind: "keep", reason: "existingPrecedence" });
      break;
    case "potentialExact":
    case "potentialPrefix":
      repairs.push({ kind: "keep", reason: "potentialConflict" });
      break;
    case "ambiguousExact":
    case "chordPrefix":
      break;
  }

  if (conflict.kind === "ambiguousExact") {
    addPreferRepair(repairs, left, right);
    addPreferRepair(repairs, right, left);
  }

  addNarrowRepair(repairs, left, right);
  addNarrowRepair(repairs, right, left);
  repairs.push({ kind: "unbind", bindingId: left.id });
  repairs.push({ kind: "unbind", bindingId: right.id });

  return { conflict: structuredClone(conflict), disposition, repairs };
}

/** Applies one explicit repair to a binding collection. No repair is selected implicitly. */
export function applyConflictRepair(
  bindings: readonly Binding[],
  repair: ConflictRepair,
): Binding[] {
  switch (repair.kind) {
    case "keep":
      return bindings.map((binding) => structuredClone(binding));
    case "unbind":
      return bindings
        .filter((binding) => binding.id !== repair.bindingId)
        .map((binding) => structuredClone(binding));
    case "prefer":
      return bindings.map((binding) =>
        binding.id === repair.bindingId
          ? { ...structuredClone(binding), priority: repair.priority }
          : structuredClone(binding),
      );
    case "narrowContext":
      return bindings.map((binding) =>
        binding.id === repair.bindingId
          ? { ...structuredClone(binding), when: structuredClone(repair.when) }
          : structuredClone(binding),
      );
  }
}

function dispositionForKind(kind: ConflictKind): ConflictDisposition {
  switch (kind) {
    case "duplicate":
      return "redundant";
    case "ambiguousExact":
      return "ambiguous";
    case "overrideExact":
      return "orderedOverride";
    case "chordPrefix":
      return "chordPrefix";
    case "potentialExact":
    case "potentialPrefix":
      return "potential";
  }
}

function addPreferRepair(
  repairs: ConflictRepair[],
  target: Binding,
  other: Binding,
): void {
  const otherPriority = other.priority ?? 0;
  if (otherPriority >= MAX_PRIORITY) return;
  repairs.push({
    kind: "prefer",
    bindingId: target.id,
    overBindingId: other.id,
    priority: otherPriority + 1,
  });
}

function addNarrowRepair(
  repairs: ConflictRepair[],
  target: Binding,
  other: Binding,
): void {
  const otherWhen = other.when ?? ALWAYS;
  if (otherWhen.op === "always") return;
  const targetWhen = target.when ?? ALWAYS;
  const exclusion: WhenExpr = { op: "not", expr: structuredClone(otherWhen) };
  const when: WhenExpr = targetWhen.op === "always"
    ? exclusion
    : { op: "all", exprs: [structuredClone(targetWhen), exclusion] };
  repairs.push({
    kind: "narrowContext",
    bindingId: target.id,
    againstBindingId: other.id,
    when,
  });
}

File: crates/input-bindings-core/src/repair.rs
use serde::{Deserialize, Serialize};

use crate::{Binding, Conflict, ConflictKind, WhenExpr};

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ConflictDisposition {
    Redundant,
    Ambiguous,
    OrderedOverride,
    ChordPrefix,
    Potential,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ConflictRepairKeepReason {
    ExistingPrecedence,
    PotentialConflict,
    RedundantSameAction,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum ConflictRepair {
    Keep {
        reason: ConflictRepairKeepReason,
    },
    Unbind {
        #[serde(rename = "bindingId")]
        binding_id: String,
    },
    Prefer {
        #[serde(rename = "bindingId")]
        binding_id: String,
        #[serde(rename = "overBindingId")]
        over_binding_id: String,
        priority: i32,
    },
    NarrowContext {
        #[serde(rename = "bindingId")]
        binding_id: String,
        #[serde(rename = "againstBindingId")]
        against_binding_id: String,
        when: WhenExpr,
    },
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConflictRepairPlan {
    pub conflict: Conflict,
    pub disposition: ConflictDisposition,
    pub repairs: Vec<ConflictRepair>,
}

pub fn plan_conflict_repairs(bindings: &[Binding], conflict: &Conflict) -> ConflictRepairPlan {
    let disposition = disposition_for_kind(&conflict.kind);
    let Some(left) = bindings
        .iter()
        .find(|binding| binding.id == conflict.left_binding_id)
    else {
        return ConflictRepairPlan {
            conflict: conflict.clone(),
            disposition,
            repairs: Vec::new(),
        };
    };
    let Some(right) = bindings
        .iter()
        .find(|binding| binding.id == conflict.right_binding_id)
    else {
        return ConflictRepairPlan {
            conflict: conflict.clone(),
            disposition,
            repairs: Vec::new(),
        };
    };

    let mut repairs = Vec::new();
    match conflict.kind {
        ConflictKind::Duplicate => repairs.push(ConflictRepair::Keep {
            reason: ConflictRepairKeepReason::RedundantSameAction,
        }),
        ConflictKind::OverrideExact => repairs.push(ConflictRepair::Keep {
            reason: ConflictRepairKeepReason::ExistingPrecedence,
        }),
        ConflictKind::PotentialExact | ConflictKind::PotentialPrefix => {
            repairs.push(ConflictRepair::Keep {
                reason: ConflictRepairKeepReason::PotentialConflict,
            });
        }
        ConflictKind::AmbiguousExact | ConflictKind::ChordPrefix => {}
    }

    if conflict.kind == ConflictKind::AmbiguousExact {
        add_prefer_repair(&mut repairs, left, right);
        add_prefer_repair(&mut repairs, right, left);
    }

    add_narrow_repair(&mut repairs, left, right);
    add_narrow_repair(&mut repairs, right, left);
    repairs.push(ConflictRepair::Unbind {
        binding_id: left.id.clone(),
    });
    repairs.push(ConflictRepair::Unbind {
        binding_id: right.id.clone(),
    });

    ConflictRepairPlan {
        conflict: conflict.clone(),
        disposition,
        repairs,
    }
}

pub fn apply_conflict_repair(bindings: &[Binding], repair: &ConflictRepair) -> Vec<Binding> {
    match repair {
        ConflictRepair::Keep { .. } => bindings.to_vec(),
        ConflictRepair::Unbind { binding_id } => bindings
            .iter()
            .filter(|binding| binding.id != *binding_id)
            .cloned()
            .collect(),
        ConflictRepair::Prefer {
            binding_id,
            priority,
            ..
        } => bindings
            .iter()
            .cloned()
            .map(|mut binding| {
                if binding.id == *binding_id {
                    binding.priority = *priority;
                }
                binding
            })
            .collect(),
        ConflictRepair::NarrowContext {
            binding_id, when, ..
        } => bindings
            .iter()
            .cloned()
            .map(|mut binding| {
                if binding.id == *binding_id {
                    binding.when = when.clone();
                }
                binding
            })
            .collect(),
    }
}

fn disposition_for_kind(kind: &ConflictKind) -> ConflictDisposition {
    match kind {
        ConflictKind::Duplicate => ConflictDisposition::Redundant,
        ConflictKind::AmbiguousExact => ConflictDisposition::Ambiguous,
        ConflictKind::OverrideExact => ConflictDisposition::OrderedOverride,
        ConflictKind::ChordPrefix => ConflictDisposition::ChordPrefix,
        ConflictKind::PotentialExact | ConflictKind::PotentialPrefix => {
            ConflictDisposition::Potential
        }
    }
}

fn add_prefer_repair(repairs: &mut Vec<ConflictRepair>, target: &Binding, other: &Binding) {
    let Some(priority) = other.priority.checked_add(1) else {
        return;
    };
    repairs.push(ConflictRepair::Prefer {
        binding_id: target.id.clone(),
        over_binding_id: other.id.clone(),
        priority,
    });
}

fn add_narrow_repair(repairs: &mut Vec<ConflictRepair>, target: &Binding, other: &Binding) {
    if other.when == WhenExpr::Always || target.when == other.when {
        return;
    }
    let exclusion = WhenExpr::Not {
        expr: Box::new(other.when.clone()),
    };
    let when = if target.when == WhenExpr::Always {
        exclusion
    } else {
        WhenExpr::All {
            exprs: vec![target.when.clone(), exclusion],
        }
    };
    repairs.push(ConflictRepair::NarrowContext {
        binding_id: target.id.clone(),
        against_binding_id: other.id.clone(),
        when,
    });
}
